### DISCUSSION TITLE
Should a uv_thread_tryjoin_np be added to detect if the thread exits?

### DISCUSSION THREAD

**Discussion by number201724** [libuv/libuv#3776] [Ideas]
Should a uv_thread_tryjoin_np be added to detect if the thread exits?

uv_thread_join is blocking, is it necessary to add a non-blocking function? For example using multiple threads and then a loop per thread and checking if the other thread ends.

**Comment by bnoordhuis** [comment]
How would you implement it on platforms that don't have a `pthread_tryjoin_np()` equivalent?

**Reply by number201724** [reply]
> How would you implement it on platforms that don't have a `pthread_tryjoin_np()` equivalent?

Return ENOSYS like uv_loop_fork? windows, linux, FreeBSD have similar functions.

**Reply by bnoordhuis** [reply]
I suppose that's an option but libuv usually refrains from adding APIs that can't be implemented on a majority of supported platforms. No openbsd/netbsd/solaris support is one thing but no macos support makes it rather less useful.

That said, you're welcome to open a pull request and see what other maintainers say. I'm not strongly opposed, just not exactly thrilled.

**Reply by number201724** [reply]
It seems that pthread_timedjoin_np can only use CLOCK_REALTIME, which is affected by NTP and seems less reliable.
It seems that pthread_peekjoin_np was added in 13.0 on FreeBSD. If this function is added, it will not work on older version.
Looks like it's not that easy to make generic on *nix. :|

**Reply by jamsilva** [reply]
tryjoin could be implemented by adding 1 level of indirection.
Instead of starting the thread with the user-provided function and argument, use a wrapper function and a struct {user_arg; completed; }.
The wrapper function simply calls the user-provided function with user_arg and sets completed afterwards.
In the tryjoin caller, test if completed is set and, if so, call pthread_join().

**Reply by vtjnash** [reply]
That can be implemented with pthread_key_create without the extra thread (it is how Julia implements this). Not entirely sure if that is desirable though, as I don't know the use case for this.
